### PR TITLE
[v5.0.1] ErgoScripit compiler without AOT costing

### PR DESCRIPTION
**Pull request by aslesarenko**
In this PR ErgoScript compiler (SigmaCompiler class) is extended with alternative implementation of source -> ErgoTree compilation which don't use AOT costing classes.

The new implementation is in GraphBuilding trait, which is the majority of the code added in this PR.

The old and new implementation of SigmaCompiler are executed speculatively and both results are saved in CompilerResult instance. This allows equivalence testing of the two implementations (like in compileWithCosting method).

Note, the tests remain unchanged which is evidence of equivalence of the new implementation.

The goal of this PR is to allow AOT code removal in subsequent PRs by switching to new ErgoScript compiler.

### CODE REVIEW

**Review by kushti — CHANGES_REQUESTED** [review]

File: sigmastate/src/main/scala/org/ergoplatform/ErgoScriptPredef.scala [hunk 1]
```diff
@@ -19,9 +19,9 @@ object ErgoScriptPredef {
 
   def compileWithCosting(env: ScriptEnv, code: String, networkPrefix: NetworkPrefix)(implicit IR: IRContext): Value[SType] = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoScriptPredef.scala:20]: fixed

File: sigmastate/src/main/scala/org/ergoplatform/ErgoScriptPredef.scala [hunk 2]
```diff
@@ -19,9 +19,9 @@ object ErgoScriptPredef {
...
     val compiler = new SigmaCompiler(networkPrefix)
-    val interProp = compiler.typecheck(env, code)
-    val IR.Pair(calcF, _) = IR.doCosting(env, interProp)
-    IR.buildTree(calcF)
+    val res = compiler.compile(env, code)
+    assert(res.calcTree == res.buildTree)
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoScriptPredef.scala:23]: moved to .compile()

File: sigmastate/src/main/scala/sigmastate/eval/GraphBuilding.scala [hunk 3]
```diff
@@ -0,0 +1,762 @@
...
+//      val env = envVals.mapValues(v => buildNode(ctxC, Map.empty, v))
+      val res = asRep[T](buildNode(ctxC, Map.empty, typed))
+      res
+    }))
+  }
+
+  /** Helper type synonym used internally */
+  protected type CompilingEnv = Map[Any, Ref[_]]
+
+  protected def buildNode[T <: SType](ctx: Ref[Context], env: CompilingEnv, node: Value[T]): Ref[T#WrappedType] = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/eval/GraphBuilding.scala:55]: added

File: sigmastate/src/main/scala/sigmastate/lang/SigmaCompiler.scala [hunk 4]
```diff
@@ -67,10 +86,20 @@ class SigmaCompiler(settings: CompilerSettings) {
...
+  /** Compiles the given ErgoScript source code. */
+  def compile(env: ScriptEnv, code: String)(implicit IR: IRContext): CompilerResult[IR.type] = {
+    val typed = typecheck(env, code)
+    val res = compileTyped(env, typed)
+    res.copy(code = code)
+  }
+
+  /** Compiles the given typed expression. */
+  def compileTyped(env: ScriptEnv, typedExpr: SValue)(implicit IR: IRContext): CompilerResult[IR.type] = {
+    val IR.Pair(calcF, costF) = IR.doCosting(env, typedExpr, true)
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/SigmaCompiler.scala:98]: fixed

File: sigmastate/src/main/scala/sigmastate/eval/TreeBuilding.scala [hunk 5]
```diff
@@ -257,8 +257,8 @@ trait TreeBuilding extends RuntimeCosting { IR: IRContext =>
         val specMethod = method.withConcreteTypes(typeSubst + (SCollection.tIV -> colTpe.elemType))
         builder.mkMethodCall(col, specMethod, args.toIndexedSeq, Map())
 
-      case BoxM.value(box) =>
-        mkExtractAmount(recurse[SBox.type](box))
+      case BoxM.value(In(box)) =>
+        mkExtractAmount(box.asBox)
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/eval/TreeBuilding.scala:261]: This change is to unify the case with the other cases. 
The `In` pattern is equivalent to `recurse` function.

**Review by kushti — APPROVED** [review]

**Pull request merged**